src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -4753,30 +4753,39 @@ BoundExpression bindSpreadElement(SpreadElementSyntax syntax, BindingDiagnosticB
                     return new BoundCollectionExpressionSpreadElement(
                         syntax,
                         expression,
+                        expressionPlaceholder: null,
+                        conversion: null,
                         enumeratorInfoOpt: null,
+                        lengthOrCount: null,
                         elementPlaceholder: null,
-                        addElementPlaceholder: null,
-                        addMethodInvocation: null,
-                        type: CreateErrorType(),
+                        iteratorBody: null,
                         hasErrors);
                 }
 
+                Debug.Assert(expression.Type is { });
+
+                var expressionPlaceholder = new BoundCollectionExpressionSpreadExpressionPlaceholder(syntax.Expression, expression.Type);
                 var enumeratorInfo = builder.Build(location: default);
                 var collectionType = enumeratorInfo.CollectionType;
                 var useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
                 var conversion = Conversions.ClassifyConversionFromExpression(expression, collectionType, isChecked: CheckOverflowAtRuntime, ref useSiteInfo);
                 Debug.Assert(conversion.IsValid);
                 diagnostics.Add(syntax.Expression, useSiteInfo);
-                expression = ConvertForEachCollection(expression, conversion, collectionType, diagnostics);
-                var elementPlaceholder = new BoundValuePlaceholder(syntax.Expression, enumeratorInfo.ElementType);
+                var convertedExpression = ConvertForEachCollection(expressionPlaceholder, conversion, collectionType, diagnostics);
+                BoundExpression? lengthOrCount;
+                if (!TryBindLengthOrCount(syntax.Expression, expressionPlaceholder, out lengthOrCount, diagnostics))
+                {
+                    lengthOrCount = null;
```

Comment:
We probably need a temp diagnostics bag that we add to `diagnostics` only if we're in the true path here. It looks like this will add use-site diagnostics from `Length` or `Count`, if any are found, even if it then returns false some some other reason.

Reply:
Good catch. I've added a test, `SpreadElement_LengthUseSiteError`, and I think we probably want to report the use-site errors (as it does now) even if `TryBindLengthOrCount` returns `false` because the reason it returned `false` might be due to the same issue that resulted in the use-site error. This is consistent with how `TryBindLengthOrCount` is used for patterns.